File: matlab/problem2/runECG3_solution.m
% -------------------------------------------------------------------------
% real data, 2 blocks, 3 features * 2 time points, classification
% -------------------------------------------------------------------------

clear
close all

% sampling frequency is 512 Hz
srate = 512;

% load filters
load('filterHP_0.5_1.5.mat')
load('filterLP_3_10.mat')


%% analyze ----------------------------------------------------------------

% phase locking pack size
packSize = 11;

% we want to discriminate blocks 2 and 7
useBlocks = [2 7];

% initialize feature structure
Features = struct('PhaseHP', [], 'PhaseBP', [], 'PLVHP', []);

for B=1:2
    
    % load data, get signal from channel 3
    load(['block_' num2str(useBlocks(B))])
    signal = block_data(:, 3);
    
    % create high-pass and band-pass filtered signals by combining filters
    signalHP = filtfilt(filterHP, 1, signal);
    signalBP = filtfilt(filterLP, 1, signalHP);
    
    % extract QRS complex positions, using the high-pass filtered signal
    [H2, HDR, s] = qrsdetect(signalHP, srate, 1);
    peakPos = H2.EVENT.POS;
    
    % compute phases
    phaseHP = angle(hilbert(signalHP));
    phaseBP = angle(hilbert(signalBP));
    
    % isolate phases
    timeWindow = round(-0.2*srate)+1:round(0.6*srate);
    phasesHP = nan(length(timeWindow), length(peakPos)-2);
    phasesBP = nan(length(timeWindow), length(peakPos)-2);
    for iEpoch=2:(length(peakPos)-1)
        tmp = unwrap(phaseHP(peakPos(iEpoch)+timeWindow));
        phasesHP(:, iEpoch-1) = tmp - tmp(timeWindow==0);
        tmp = unwrap(phaseBP(peakPos(iEpoch)+timeWindow));
        phasesBP(:, iEpoch-1) = tmp - tmp(timeWindow==0);
    end
    
    % compute phase locking values
    margin = floor(packSize/2);
    plvHP = nan(length(timeWindow), size(phasesHP, 2));
    for iEpoch=margin+1:size(plvHP, 2)-margin
        thisPack = iEpoch + (-margin:margin);
        plvHP(:, iEpoch) = abs(mean(exp(1i*phasesHP(:, thisPack)), 2));
    end
    
    % store features
    Features(B).PhaseHP = phasesHP(:, margin+1:end-margin);
    Features(B).PhaseBP = phasesBP(:, margin+1:end-margin);
    Features(B).PLVHP = plvHP(:, margin+1:end-margin);
    
end


%% classify ---------------------------------------------------------------

% define sample time points
t1 = round((0.2 + 0.03)*srate);
t2 = round((0.2 + 0.29)*srate);

% construct feature matrix for each block
nEpochs = min(size(Features(1).PhaseHP, 2), size(Features(2).PhaseHP, 2));
X = nan(2*nEpochs, 6);
for B=1:2
    rows = (B-1)*nEpochs + (1:nEpochs);
    X(rows, 1) = Features(B).PhaseHP(t1, 1:nEpochs);
    X(rows, 2) = Features(B).PhaseHP(t2, 1:nEpochs);
    X(rows, 3) = Features(B).PhaseBP(t1, 1:nEpochs);
    X(rows, 4) = Features(B).PhaseBP(t2, 1:nEpochs);
    X(rows, 5) = Features(B).PLVHP(t1, 1:nEpochs);
    X(rows, 6) = Features(B).PLVHP(t2, 1:nEpochs);
end

% train and validate logit model, linear SVM, and RBF SVM
kCross = 10;
L = [ones(nEpochs, 1); zeros(nEpochs, 1)];
modelType = {'logreg', 'linsvm', 'rbfsvm'};
for iModel=1:3
    pCorrect = modelFitVal2_solution(X, L, kCross, modelType{iModel});
    fprintf('\nPerformance %6s: %3i %%\n', ...
        modelType{iModel}, round(100*pCorrect));
end

File: python/problem2/runECG3_solution.py
"""
real data, 2 blocks, 3 features * 2 time points, classification

"""

from pylab import *
from scipy.io import loadmat
import scipy.signal as sig

from modelFitVal2 import modelFitVal2
from rpeakdetect import detect_beats

# sampling frequency is 512 Hz
f_s = 512

# load filters designed with pyfda
filterHP = load('highpass_filter.npz')
filterLP = load('lowpass_filter.npz')

# analyze

# phase locking pack size
packSize = 25

# we want to discriminate blocks 2 and 7
useBlocks = [2, 7]

Features = {}

for B in [0, 1]:
    print('Process block', useBlocks[B])

    # load data, get signal from channel 3
    block_data = loadmat('../../data/ECG/block_' + str(useBlocks[B]))
    block_data = block_data['block_data'].T
    ecg_signal = block_data[2]
    ecg_signal = ecg_signal.astype('float64')

    # create high-pass and band-pass filtered signals by combining filters
    b, a = filterHP['ba']
    signalHP = sig.filtfilt(b, 1.0, ecg_signal)
    b, a = filterLP['ba']
    signalBP = sig.filtfilt(b, 1.0, signalHP)

    # extract QRS complex positions, using the high-pass filtered signal
    peakPos = detect_beats(signalHP, f_s)

    # compute phases by getting the angle of the analytical signals
    phaseHP = angle(sig.hilbert(signalHP))
    phaseBP = angle(sig.hilbert(signalBP))

    # isolate phases. define a time window around each QRS peak, cut out
    # the single phase epochs (unwrapped high-pass and band-pass filtered
    # phases), align each epoch to its value at the time of the QRS peak,
    # and stack them into matrices. leave out the first and last epoch to
    # avoid indexing errors
    timeWindow = arange(round(-0.2 * f_s), round(0.6 * f_s))
    windows = peakPos[1:-1].reshape(-1, 1) + timeWindow
    phasesHP = unwrap(phaseHP[windows])
    phasesHP -= phasesHP[:, timeWindow == 0]
    phasesBP = unwrap(phaseBP[windows])
    phasesBP -= phasesBP[:, timeWindow == 0]

    # compute phase locking values over time
    packNumber = len(phasesHP) // packSize
    packs = arange(packNumber * packSize)
    packs = reshape(packs, [packNumber, packSize])
    plvHP = phasesHP[packs]
    plvHP = abs(mean(exp(1j * plvHP), axis=1))

    Features[B] = {
        'PhaseHP': phasesHP,
        'PhaseBP': phasesBP,
        'PLVHP': plvHP,
    }


# ---- classify ---- #

# specify epochs to use. (number of epochs should be a multiple of
# packSize)
useEpochs = range(
    packSize * min(len(Features[0]['PLVHP']),
                   len(Features[1]['PLVHP'])))
nEpochs = len(useEpochs)

# define 2 sample time points around the QRS complex and the T-wave,
# respectively. note that at exactly 200 ms phases are set to zero!
t1 = round((0.2 + 0.03) * f_s)
t2 = round((0.2 + 0.29) * f_s)

print('construct feature matrix')  # for each block
featureMatrix = {}
for B in [0, 1]:
    feat = empty([nEpochs, 6])
    feat[:, 0] = Features[B]['PhaseHP'][useEpochs, t1]
    feat[:, 1] = Features[B]['PhaseHP'][useEpochs, t2]
    feat[:, 2] = Features[B]['PhaseBP'][useEpochs, t1]
    feat[:, 3] = Features[B]['PhaseBP'][useEpochs, t2]
    tmp = Features[B]['PLVHP'][:, t1]
    print('tmpshape1', tmp.shape)
    tmp = repeat(tmp, packSize)
    print('tmpshape2', tmp.shape)
    feat[:, 4] = tmp[useEpochs]
    tmp = Features[B]['PLVHP'][:, t2]
    tmp = repeat(tmp, packSize)
    feat[:, 5] = tmp[useEpochs]
    featureMatrix[B] = feat


# set number of cross-validation cycles
kCross = 5

# train and validate logit model, linear SVM, and RBF SVM
X = vstack([featureMatrix[0], featureMatrix[1]])
L = vstack([ones([nEpochs, 1]), zeros([nEpochs, 1])])
print(X.shape, L.shape)
for modelType in ['logit', 'linsvm', 'rbfsvm']:
    pCorrect = modelFitVal2(X, L, kCross, modelType)
    print('\nPerformance {}: {}%\n'.format(
        modelType,
        round(100 * pCorrect)))
